c10/util/intrusive_ptr.h

```diff
@@ -277,27 +278,62 @@ class intrusive_ptr final {
     }
   }
 
+  C10_NOINLINE void reset_zero_refcount_slow_path() noexcept {
+    reset_zero_refcount();
+  }
+
+  void reset_zero_refcount() noexcept {
+    // See comment above about weakcount. As long as refcount>0,
+    // weakcount is one larger than the actual number of weak references.
+    // So we need to decrement it here.
+    bool should_delete =
+        target_->weakcount_.load(std::memory_order_acquire) == 1;
+    if (!should_delete) {
+      // justification for const_cast: release_resources is basically a
+      // destructor and a destructor always mutates the object, even for const
+      // objects. NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+      const_cast<std::remove_const_t<TTarget>*>(target_)->release_resources();
+      should_delete =
+          detail::atomic_weakcount_decrement(target_->weakcount_) == 0;
+    }
+    if (should_delete) {
+      delete target_;
+    }
+  }
+
+  #pragma GCC diagnostic push
+  #pragma GCC diagnostic ignored "-Wstrict-aliasing"
```

Comment:
There are `C10_DIAGNOSTIC_PUSH_AND_IGNORED_IF_DEFINED`, which probably be more appropriate here
https://github.com/pytorch/pytorch/blob/4a49e2b52d90a0f05fc7ead1aeb3fbeab014a2ad/c10/macros/Macros.h#L540